draft/0002-subdocapi.md

```diff
@@ -0,0 +1,351 @@
+within a document, thus a valid path may be:
+
+    foo.bar.baz
+
+or
+
+    `foo`.`bar`.`baz`
+
+or
+
+    `foo.some_array[0]`
+
+or
+
+    `foo.some_array[-1]`
+
+As in N1QL, special path characters may be escaped via a backtick (`` ` ``).
+
+Some operations (notably, those taking a path to an array) may accept the
+_empty path_ which refers to the entire document. This is valid if the document
+is a bare JSON array.
+
+Each path component must be valid JSON (without the enclosing quotes). This
+means that things like tabs, quotes, and backslashes must be escaped. To think
+of it another way, the path matching algorithm in the server works via simple
+string comparison, so a JSON key `"bs\\ and \t tabs are here" would be matched
+by a path component of `"bs\\ and \t tabs are here"`.
+
+The SDK MUST verify that a path is non-empty for operations which do not
+support empty paths. All other forms of path verification is done on the server
+(but an empty path for unsupported operations will return a memcached `EINVAL`
+which isn't really helpful).
+
+## Single-path Methods
+
+Methods are named exactly like their top-level counterparts, and should also
+exist at the top level as well.
+
+Error codes are common to most operations.
+
+### `get_in(docid, path)`
+Gets a single path from the document. Returns a MetaDocFragment containing the
+path value.
+
+* `docid`: Document ID
+* `path`: The sub-document path
+
+### `exists_in(docid, path)`
+
+Like `get_in`, but does not return the value (only checks if it exists)`
+
+### `upsert_in(docid, path, value, create_parents, cas, ttl, persist_to, replicate_to)`
+Unconditionally places the specified value in the provided path.
+
+* `docid`: Document ID
+* `path`: the sub-document path
+* `value`: the sub-document value. This should be anything JSON-serializable.
+  JSON primitives such as true, false, null, strings and numbers are accepted
+  too.
+* `create_parents`: Boolean parameter indicating whether the full parent tree
+  should be created. By default, only the _last_ path element is created if
+  it does not exist.
+* `cas`: Exactly the same meaning as it has in top-level CRUD operations.
+* `ttl`: Expiration time. This has the same meaning as it does in the CRUD
+  operations. The TTL can be set for _all_ mutation methods.
+* `persist_to`, `replicate_to`: Durability requirements. Functions exactly as in CRUD
+
+### `insert_in(docid, path, value, create_parents, cas, ttl, persist_to, replicate_to)`
+Exactly like `upsert_in`, except the path must not already exist.
+If it does, `PathExists` is thrown.
+
+### `replace_in(docid, path, value, cas, ttl, persist_to, replicate_to)`
+Exactly like `upsert_in`, except the penultimate path must exist. If it 
+not exist, `PathNotFound` is thrown. Unlike `insert_in` and `upsert_in`,
+`replace_in` can work on both dictionary and array values
+
+### `extend_in(docid, path, value_or_values, direction, create_parents, cas, ttl, persist_to, replicate_to)`
```

Comment:
Because those names are taken to mean string append/prepend at the top-level. I think it's important not to conflate the two terms (even if it's unfortunate that `append` has primacy over `append_in`). This is still up for discussion though.